### PR TITLE
random edge reorientation

### PR DESCRIPTION
**Pull request by sbromberger**

### CODE REVIEW

**Comment by sbromberger** [comment]
cc @simonschoelly

File: src/SimpleGraphs/generators/randgraphs.jl [hunk 1]
```diff
@@ -78,6 +78,44 @@ end
...
+"""
+	SimpleDiGraph(g, p_in, p_out; seed=-1)
+
+Create a directed graph from an undirected graph `g` where the direction of each
+edge from `u` to `v` (`u < v`) in `g` is added as a directed edge from `u` to `v`
+with probability `p_out`, from `v` to `u` with probability `p_in`, or in both
+directions with probability `1 - p_in - p_out`. If `p_in + p_out ≥ 1`, `p_out`
+will be truncated to `1 - p_in`, and no bidirectional edges will be created.
+"""
+function SimpleDiGraph(g::SimpleGraph, _p_in::Real, _p_out::Real; seed::Int=-1)
```
Comment (simonschoelly) [inline src/SimpleGraphs/generators/randgraphs.jl]: I am still not convinced, that a constructor is the right thing to use here. It does not give any hints to someone reading the code using this, what this function does. Furthermore, I don't know how much this function will be used, so it will make it impossible to overload this method signature in the future.

If you still want to use the constructor, then I think it should at least redirect to another function, like we do for erdos_renyi.

File: src/SimpleGraphs/generators/randgraphs.jl [hunk 2]
```diff
@@ -78,6 +78,44 @@ end
...
 
+"""
+	SimpleDiGraph(g, p_in, p_out; seed=-1)
+
+Create a directed graph from an undirected graph `g` where the direction of each
+edge from `u` to `v` (`u < v`) in `g` is added as a directed edge from `u` to `v`
+with probability `p_out`, from `v` to `u` with probability `p_in`, or in both
+directions with probability `1 - p_in - p_out`. If `p_in + p_out ≥ 1`, `p_out`
+will be truncated to `1 - p_in`, and no bidirectional edges will be created.
+"""
```
Comment (simonschoelly) [inline src/SimpleGraphs/generators/randgraphs.jl]: It would be helpful, if you specified here, that the order is taken by viewing the vertices as integers, otherwise it is not that easy to comprehend I guess.

File: src/SimpleGraphs/generators/randgraphs.jl [hunk 3]
```diff
@@ -78,6 +78,44 @@ end
...
+function SimpleDiGraph(g::SimpleGraph, _p_in::Real, _p_out::Real; seed::Int=-1)
+    p_in, p_out = promote(_p_in, _p_out)
+    T = typeof(p_in)
+    edgelist = Vector{edgetype(g)}()
+    ne_g = ne(g)
+    p_both = max(one(T) - p_in - p_out, zero(T))
+    thresh_in = (zero(T), p_in)
+    thresh_out = (p_in, p_in + p_out)
+    sizehint!(edgelist, ne_g + Int(ceil(ne_g * p_both)))
```
Comment (simonschoelly) [inline src/SimpleGraphs/generators/randgraphs.jl]: Maybe that can be optimized by doing something like 
```
sizehint!(edgelist, ne_g + Int(ceil(  1.1 * ne_g * p_both)))
```
Otherwise, in half of the cases, the `sizehint!` underestimates the necessary space and then Julia will double the space for `edgelist`.

File: test/simplegraphs/generators/randgraphs.jl [hunk 4]
```diff
@@ -15,6 +15,21 @@
...
+        d1 = @inferred(DiGraph(g, 0.0, 1.0))
+        @test d1 == reverse(dg10)
+        d2 = @inferred(DiGraph(g, 2//7, 5//9))
+        @test ne(d2) >= ne(g)
+        d3 = @inferred(DiGraph(g, 0, 0))
+        @test d3 == DiGraph(g)
+        d4 = @inferred(DiGraph(g, -1, 200))
+        @test d4 == reverse(dg10)
+    end
+
```
Comment (simonschoelly) [inline test/simplegraphs/generators/randgraphs.jl]: Maybe add a test, that verifies, that we get the same result when using the same seed.
Also some tests for edge cases, such as the zero graph, graphs without edges and graphs with self-loops.

**Comment by simonschoelly** [comment]
Code looks code to me, I am not entirely conviced that the interace is easily understandable yet.
Can we still have a function `random_orientation(g, propbabiliy_bidirectional)`, that uses your function then? This would fit nicely together with the function `random_orintation_dag` and is easy to understand.

File: src/SimpleGraphs/generators/randgraphs.jl (hunk 1 above)
Comment (sbromberger) [reply src/SimpleGraphs/generators/randgraphs.jl]: I'm ok with calling it something else. What do you suggest?

File: src/SimpleGraphs/generators/randgraphs.jl (hunk 3 above)
Comment (sbromberger) [reply src/SimpleGraphs/generators/randgraphs.jl]: Good point. I'm going to be even more conservative and use 1.2.

File: src/SimpleGraphs/generators/randgraphs.jl (hunk 2 above)
Comment (sbromberger) [reply src/SimpleGraphs/generators/randgraphs.jl]: What else could they be? `SimpleGraphs` requires integer vertices (it's in the documentation).

File: test/simplegraphs/generators/randgraphs.jl (hunk 4 above)
Comment (sbromberger) [reply test/simplegraphs/generators/randgraphs.jl]: I'm good with all of that except for the self-loops. We're not really supporting them (they're definitely second-class).
